ambry-network/src/main/java/com/github/ambry/network/http2/Http2BlockingChannel.java

```diff
@@ -68,139 +45,119 @@
-        p.addLast(new Http2StreamFrameToHttpObjectCodec(false));
-        p.addLast(new HttpObjectAggregator(1024 * 1024));
-        p.addLast(new SimpleChannelInboundHandler<FullHttpResponse>() {
-          @Override
-          protected void channelRead0(ChannelHandlerContext ctx, FullHttpResponse msg) throws Exception {
-            Integer streamId = msg.headers().getInt(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text());
-            if (streamId == null) {
-              logger.error("Http2ResponseHandler unexpected message received: " + msg);
-              return;
-            }
-            ctx.channel().attr(RESPONSE_PROMISE).getAndSet(null).setSuccess(msg.content().retainedDuplicate());
-            logger.trace("Stream response received.");
-          }
-        });
-      }
-    });
   }
 
   @Override
   public void disconnect() throws IOException {
-    channel.disconnect().syncUninterruptibly();
-    workerGroup.shutdownGracefully();
+
   }
 
+  /**
+   * A new stream channel is created for each send.
+   */
   @Override
   public void send(Send request) throws IOException {
-    ByteBufferChannel byteBufferChannel = new ByteBufferChannel(ByteBuffer.allocate((int) request.sizeInBytes()));
-    while (!request.isSendComplete()) {
-      request.writeTo(byteBufferChannel);
+    Channel streamChannel;
+    try {
+      streamChannel = http2MultiplexedChannelPool.acquire()
+          .get(http2ClientConfig.http2BlockingChannelAcquireTimeoutMs, TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
+      throw new IOException("Can't acquire stream channel from " + getRemoteHost() + ":" + getRemotePort(), e);
     }
-    byteBufferChannel.getBuffer().position(0);
-    ByteBuf byteBuf = Unpooled.wrappedBuffer(byteBufferChannel.getBuffer());
-
-    Http2StreamChannel childChannel = http2StreamChannelBootstrap.open().syncUninterruptibly().getNow();
-    Http2Headers http2Headers = new DefaultHttp2Headers().method(HttpMethod.POST.asciiName()).scheme("https").path("/");
-    responsePromise = childChannel.eventLoop().newPromise();
-    childChannel.attr(RESPONSE_PROMISE).set(responsePromise);
-
-    DefaultHttp2HeadersFrame headersFrame = new DefaultHttp2HeadersFrame(http2Headers, false);
-    DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(byteBuf, true);
-    childChannel.write(headersFrame);
-    childChannel.write(dataFrame);
-    childChannel.flush();
+
+    ChannelPipeline p = streamChannel.pipeline();
+    p.addLast(new Http2ClientStreamStatsHandler(http2ClientMetrics));
+    p.addLast(new Http2StreamFrameToHttpObjectCodec(false));
+    p.addLast(new HttpObjectAggregator(http2ClientConfig.http2MaxContentLength));
+    p.addLast(new SimpleChannelInboundHandler<FullHttpResponse>() {
+      @Override
+      protected void channelRead0(ChannelHandlerContext ctx, FullHttpResponse msg) throws Exception {
+        ctx.channel().attr(RESPONSE_PROMISE).getAndSet(null).setSuccess(msg.content().retainedDuplicate());
```

Comment:
this should also have a null check in case exceptionCaught called getAndSet first